crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -1259,23 +1148,60 @@ impl actor::HcP2p for HolochainP2pActor {
 
     fn send_validation_receipts(
         &self,
-        _dna_hash: DnaHash,
-        _to_agent: AgentPubKey,
-        _receipts: ValidationReceiptBundle,
+        dna_hash: DnaHash,
+        to_agent: AgentPubKey,
+        receipts: ValidationReceiptBundle,
     ) -> BoxFut<'_, HolochainP2pResult<()>> {
-        Box::pin(async move { todo!() })
-    }
+        Box::pin(async move {
+            let kitsune = self.kitsune()?;
+            let space_id = dna_hash.to_k2_space();
+            let space = kitsune.space(space_id.clone()).await?;
 
-    fn new_integrated_data(&self, _dna_hash: DnaHash) -> BoxFut<'_, HolochainP2pResult<()>> {
-        Box::pin(async move { todo!() })
+            let to_url = match space
+                .peer_store()
+                .get(to_agent.to_k2_agent())
+                .await?
+                .and_then(|i| i.url.clone())
+            {
+                Some(to_url) => to_url,
+                None => {
+                    return Err(HolochainP2pError::other(
+                        "send_validation_receipts could not find url for peer",
+                    ))
+                }
+            };
+
+            let req = crate::wire::WireMessage::validation_receipts(to_agent.clone(), receipts);
+
+            let start = std::time::Instant::now();
+
+            let out = self.send_notify(&space, to_url, req).await;
+
+            timing_trace_out!(out, start, a = "send_validation_receipts");
+
+            Ok(())
+        })
     }
 
     fn authority_for_hash(
         &self,
-        _dna_hash: DnaHash,
-        _basis: OpBasis,
+        dna_hash: DnaHash,
+        basis: OpBasis,
     ) -> BoxFut<'_, HolochainP2pResult<bool>> {
-        Box::pin(async move { todo!() })
+        Box::pin(async move {
+            let loc = basis.get_loc();
+            let kitsune = self.kitsune()?;
+            let space_id = dna_hash.to_k2_space();
+            let space = kitsune.space(space_id.clone()).await?;
+
+            for agent in space.local_agent_store().get_all().await? {
```

Comment:
This'll be true if any of the local agents is authority for a hash, so with one local agent who is an authority and another that isn't it'll be true. I wonder what the function is used for?

Reply:
Holochain uses it to decide if it needs to go out to the network for something. If any joined agent is an authority, we should have the data locally and thus not need to go out for it, even if other joined local agents are not authorities.

Reply:
That and it also uses it to decide whether authored data should be put in the DHT database. That logic will be quite screwy with the initial arc set to empty. We rely on that being FULL during genesis currently. There will be some subtle test breaks with that. 

I'm not saying that's a problem, I think the other code is a problem and not this code - just something that will come up when we try to get the tests passing